### PR TITLE
feat: Add concept of cache expiry and a subcommand to delete expired files

### PR DESCRIPTION
**Pull request by untitaker**
* CacheActor will look at a file, and if it's considered expired, will just overwrite it.
* `symbolicator cleanup` will go through all caches synchronously and delete files it considers expired.

### CODE REVIEW

File: src/actors/cache.rs [hunk 1]
```diff
@@ -91,21 +80,21 @@ impl<T: CacheItemRequest> Handler<ComputeMemoized<T>> for CacheActor<T> {
...
-            let cache_dir = self.cache_dir.clone();
+            metric!(counter(&format!("caches.{}.channel.miss", self.config.name)) += 1);
+            let config = self.config.clone();
 
             let (tx, rx) = oneshot::channel();
 
-            let file = if let Some(ref dir) = self.cache_dir {
-                NamedTempFile::new_in(dir)
+            let file = if let Some(ref dir) = config.cache_dir {
+                fs::create_dir_all(dir).and_then(|_| NamedTempFile::new_in(dir))
```
Comment (jan-auer) [inline src/actors/cache.rs]: ❤️

File: src/cache.rs [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+#[derive(Debug, Clone)]
+pub struct Cache {
+    /// Cache identifier used for metric names.
+    pub name: &'static str,
+
+    /// Directory to use for storing cache items. Will be created if it does not exist.
+    pub cache_dir: Option<PathBuf>,
+
+    /// Time when this process started.
+    start_time: SystemTime,
```
Comment (jan-auer) [inline src/cache.rs]: Maybe make this public too?

File: src/cache.rs [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+
+        // Immediately expire malformed items that have been created before this process started.
+        // See docstring of MALFORMED_MARKER
+        if MALFORMED_MARKER.len() as u64 == metadata.len() {
+            let created_at = metadata.created()?;
+            if created_at < self.start_time {
+                log::trace!("Created at is older than start time");
+                let mut file = File::open(path)?;
+                let mut buf = vec![0; MALFORMED_MARKER.len()];
+                file.read_exact(&mut buf)?;
```
Comment (jan-auer) [inline src/cache.rs]: This is excellent!

File: src/cache.rs [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+    fn check_expiry(&self, path: &Path) -> io::Result<bool> {
+        let metadata = path.metadata()?;
+
+        log::trace!("File length: {}", metadata.len());
+
+        // Immediately expire malformed items that have been created before this process started.
+        // See docstring of MALFORMED_MARKER
+        if MALFORMED_MARKER.len() as u64 == metadata.len() {
+            let created_at = metadata.created()?;
+            if created_at < self.start_time {
```
Comment (jan-auer) [inline src/cache.rs]: This should check something similar to:

```rust
if created_at < cmp::max(self.start_time, now() - config.invalid_cache_timeout)
```

File: src/cache.rs (hunk 2 above)
Comment (untitaker) [reply src/cache.rs]: Ideally users wouldn't need to care about this. Tests already have access and for those I can imagine it being useful.

File: src/cache.rs (hunk 4 above)
Comment (untitaker) [reply src/cache.rs]: Note that a run of `symbolicator cleanup` already unconditionally removes all invalid caches. This was intended behavior to achieve the same effect you want to have here (periodically clean invalid caches regardless of start time).

File: src/cache.rs (hunk 2 above)
Comment (untitaker) [reply src/cache.rs]: Fixed by making everything private.

File: src/cache.rs (hunk 4 above)
Comment (untitaker) [reply src/cache.rs]: Discussed IRL. @jan-auer wants to establish the principle that correctness should not rely on cleanup.

**Pull request merged**